pkg/providers/amifamily/ami.go

```diff
@@ -132,6 +132,38 @@ func (p *Provider) Get(ctx context.Context, nodeTemplate *v1alpha1.AWSNodeTempla
 	return amiIDs, nil
 }
 
+func (p *Provider) GetAMIWithRequirements(ctx context.Context, nodeTemplate *v1alpha1.AWSNodeTemplate, instanceTypes []*cloudprovider.InstanceType, amiFamily AMIFamily) (map[AMI]scheduling.Requirements, error) {
+	kubernetesVersion, err := p.KubeServerVersion(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("getting kubernetes version %w", err)
+	}
+	amiIDs := map[string][]*cloudprovider.InstanceType{}
+	amiRequirements, err := p.getAMIRequirements(ctx, nodeTemplate)
+	if err != nil {
+		return nil, err
+	}
+	if len(amiRequirements) == 0 {
+		for _, instanceType := range instanceTypes {
+			amiID, err := p.getDefaultAMIFromSSM(ctx, amiFamily.SSMAlias(kubernetesVersion, instanceType))
+			if err != nil {
+				return nil, err
+			}
+			amiIDs[amiID] = append(amiIDs[amiID], instanceType)
+		}
+		amifilter := map[string]string{"aws-ids": strings.Join(lo.Keys(amiIDs), ",")}
+		amiRequirements, err = p.selectAMIs(ctx, amifilter)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Iterate through AMIs in order of creation date to use latest AMI
+	amis := sortAMIsByCreationDate(amiRequirements)
+	return lo.Associate(amis, func(ami AMI) (AMI, scheduling.Requirements) {
```

Comment:
I don't think this is doing what you think it's doing. It will associate items by key. In this case it's associating by AMI ID as the key, which two AMIs shouldn't have the same key, so the association isn't doing anything.

I think what you want is to hash the requirements and associate the requirements that have the same hash value